File: day09/marble_mania.cpp
#include <stdio.h>
#include <stdlib.h>
#include <list>
#include <algorithm>

using namespace std;

int main(int argc, char *argv[]) {
    int num_players = atoi(argv[1]);
    int num_marbles = atoi(argv[2]);
    unsigned long long *score = new unsigned long long[num_players];
    list<int> circle;
    circle.push_back(0);
    auto idx = circle.begin();
    int player = 0;

    for (int i = 0; i < num_players; i++)
        score[i] = 0;
    
    for (int marble = 1; marble <= num_marbles; marble++) {
        if (marble % 23 == 0) {
            for (int i = 0; i < 6; i++)
                if (--idx == circle.begin())
                    idx = circle.end();
            idx--;
            score[player] += (marble + *idx);
            idx = circle.erase(idx);
        } else {
            if (++idx == circle.end())
                idx = circle.begin();
            idx++;
            idx = circle.insert(idx, marble);
        }
        player = (player + 1) % num_players;

        // for (auto idx2 : circle)
        //     printf("%d ", idx2);
        // printf("\n");
    }

    printf("part1: %llu\n", *max_element(score, score+num_players));
}

File: day09/marble_mania.py
#!/usr/bin/env python
from sys import argv

num_players = int(argv[1])
num_marbles = int(argv[2])

circle = [0, 1]
score = [0] * num_players
idx = 1
player = 0
for marble in range(2, num_marbles+1):
    if marble % 100000 == 0:
        print('marble =', marble)
    if marble % 23 == 0:
        idx = (idx - 7) % len(circle)
        score[player] += marble + circle.pop(idx)
    else:
        idx += 2
        if idx == len(circle):
            circle.append(marble)
        else:
            idx %= len(circle)
            circle.insert(idx, marble)
#    print(idx, circle)
    player = (player + 1) % num_players

print('part1:', max(score))
